flowredux/src/commonMain/kotlin/com/freeletics/flowredux/dsl/ChangeState.kt

```diff
@@ -35,16 +40,30 @@ public class MutateState<InputState : S, S>(
         reducer(state as InputState)
 }
 
+internal class UnsafeMutateState<InputState, S>(
+    internal val reducer: InputState.() -> S
+) : ChangeState<S>() {
+    @Suppress("UNCHECKED_CAST")
+    internal fun reduceImpl(state: S): S =
+        reducer(state as InputState)
+}
 
 /**
  * No change, this is semantically equivalent to use [OverrideState] and pass in the previous state
  */
+@Deprecated("call noChange() on State instead", level = DeprecationLevel.WARNING)
 public object NoStateChange : ChangeState<Nothing>()
 
+//TODO rename after removing deprecated NoStateChange
+internal object InternalNoStateChange : ChangeState<Nothing>()
+
 internal fun <S> ChangeState<S>.reduce(state: S): S {
```

Comment:
Not 100% sure, but if we make all these `ChangeState` implementation internal only then we cannot write simple unit tests for the "handlers" anymore, do we?

example

```kotlin
fun handleOnEnter(state : State<MyState>) : ChangeState<MyState> { ... } // I want to unit test this function

@Test
fun `test handleOnEnter`(){
   val stateSnapshot : State<MyState> = ...
   
   val expectedState : MyState = ...
   
   val actualChangeState : ChangeState<MyState> =  handleOnEnter(stateSnapshot)

  val actual : MyState = actualChangeState. ???  // How do we get the next state out of a ChangeState object ? 
  
  assertEquals( expectedState , actual)
}
```


Maybe we need to lift `reduce()` up as an `abstract fun` into `ChangeState`:


```kotlin
sealed class ChangeState<S : Any> {
   abstract fun reduce(state : S) : S
}
```

Reply:
Looks like it was already not possible to do this? We should definitely change that and I think putting reduce directly into ChangeState makes the most sense

Reply:
Will do this in a follow up

Reply:
👍 Very strange, I do remember having written some tests like that a year ago or so in another side project, just try to remember which one 😃 so that I could check it ...

But I guess I never updated that side project then ...